### PR TITLE
ass_shaper improvements extracted from threading PR

### PR DESCRIPTION
**Pull request by rcombs**
Changes to `ass_shaper` made during work on #636 which don't require threading/atomics/locks:
- Keep the `hb_buffer_t` around between calls (it turns out to be fairly expensive to free and re-create in some cases)
- Create the `hb_font_t` eagerly during `add_face`
- Some consolidation of how we access the `FT_Face` and other data during font callbacks

The later changes are required for efficient access to the font object during shaping when running multi-threaded (they allow us to use lock-free cacheing for reads, which avoids most of the locking overhead around the `FT_Face` that we'd otherwise see).

### CODE REVIEW

File: libass/ass_font.c [hunk 1]
```diff
@@ -457,8 +458,16 @@ static int add_face(ASS_FontSelector *fontsel, ASS_Font *font, uint32_t ch)
     set_font_metrics(face);
 
     font->faces[font->n_faces] = face;
-    font->faces_uid[font->n_faces++] = uid;
-    return font->n_faces - 1;
+    font->faces_uid[font->n_faces] = uid;
+    if (!ass_create_hb_font(font, font->n_faces)) {
+        FT_Done_Face(font->faces[font->n_faces]);
```
Comment (astiob) [inline libass/ass_font.c:463]: Nit: I feel this would be clearer as `FT_Done_Face(face);`

File: libass/ass_shaper.c [hunk 2]
```diff
@@ -451,54 +455,25 @@ get_contour_point(hb_font_t *font, void *font_data, hb_codepoint_t glyph,
...
-    struct ass_shaper_metrics_data *metrics =
-        font->shaper_priv->metrics_data[info->face_index];
+    update_hb_size(hb_font, font->faces[info->face_index]);
+    metrics->metrics_cache = shaper->metrics_cache;
     metrics->hash_key.font = info->font;
     metrics->hash_key.face_index = info->face_index;
     metrics->hash_key.size = info->font_size;
+    metrics->vertical = info->font->desc.vertical;
+
+    hb_font_set_funcs(hb_font, shaper->font_funcs, metrics, free);
```
Comment (astiob) [inline libass/ass_shaper.c:474]: Is it OK to take a shared `hb_font` from the `ASS_Font` and to apply a unique instance of `ass_shaper_metrics_data` to it via `hb_font_set_funcs`? Will this not interfere with threading?

**Review by astiob — COMMENTED** [review]
Thanks for these smaller PRs!

File: libass/ass_font.c (hunk 1 above)
Comment (rcombs) [reply libass/ass_font.c]: Sure; done.

File: libass/ass_font.c [hunk 3]
```diff
@@ -457,8 +458,16 @@ static int add_face(ASS_FontSelector *fontsel, ASS_Font *font, uint32_t ch)
     set_font_metrics(face);
 
     font->faces[font->n_faces] = face;
-    font->faces_uid[font->n_faces++] = uid;
-    return font->n_faces - 1;
+    font->faces_uid[font->n_faces] = uid;
+    if (!ass_create_hb_font(font, font->n_faces)) {
+        FT_Done_Face(face);
+        goto fail;
```
Comment (MrSmile) [inline libass/ass_font.c:464]: That looks unnecessary complicated. Something like
```c
    if (ass_create_hb_font(font, font->n_faces))
        return font->n_faces++;
    FT_Done_Face(face);
    return -1;
```
should be better.

File: libass/ass_shaper.c (hunk 2 above)
Comment (rcombs) [reply libass/ass_shaper.c:474]: See `ass_shaper: don't mutate the shared hb_font_t during shaping` in #636 (currently 5ad73525b8834892cf958386c8dc6bc04915d97d); I use `hb_font_create_sub_font` to create a local `hb_font_t` that references the shared one, but has its own funcs font-data.

…though while writing this, I realized I could just cache the sub-font directly and save a decent bit of time for negligible memory cost, so I'm doing that.

File: libass/ass_font.c (hunk 3 above)
Comment (rcombs) [reply libass/ass_font.c:464]: The fail path will later be needed for threading (to unlock), and I don't see a particular reason to reorganize this here only to redo it again in the threading branch.

**Review by astiob — APPROVED** [review]

**Pull request merged**